{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f6490dcf-f1ca-4764-9406-d871d332f0dd",
   "metadata": {},
   "source": [
    "# Macrocyte & Macrophage dataset\n",
    "\n",
    "This notebook is part of the paper titled, \"Single-Cell Trajectory Inference for Detecting Transient Events in Biological Processes\" by Hutton and Meyer. The data is from the 2021 Specht et al. paper titled, \"[Single-cell proteomic and transcriptomic analysis of macrophage heterogeneity using SCoPE2](https://doi.org/10.1186/s13059-021-02267-5)\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f3779e0-d4cb-425d-8511-5c77ddb5c140",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import wavelet_pseudotime as wp\n",
    "import wavelet_pseudotime.load_data\n",
    "from matplotlib import pyplot as plt\n",
    "from IPython.display import clear_output\n",
    "from importlib import reload\n",
    "from importlib import reload\n",
    "import numpy as np\n",
    "from anndata import AnnData\n",
    "from typing import Union, List, Tuple\n",
    "from wavelet_pseudotime.windowing import Window, GaussianWindow, RectWindow, ConfinedGaussianWindow\n",
    "from collections import defaultdict as dd\n",
    "from wavelet_pseudotime.wavelets import WaveletTransform, get_max_wavelets, get_max_scored_wavelets\n",
    "from typing import Literal\n",
    "import scanpy as sc\n",
    "from wavelet_pseudotime.binning import quantile_binning_anndata\n",
    "from wavelet_pseudotime.process import window_trajectory\n",
    "from scipy.ndimage import gaussian_filter1d\n",
    "import seaborn as sns\n",
    "import os\n",
    "import matplotlib.gridspec as gridspec\n",
    "\n",
    "from datetime import datetime\n",
    "date_str = datetime.now().strftime(\"%Y_%m_%d\")\n",
    "r_dir = f\"{date_str}_slavov\"  # results directory\n",
    "if not os.path.exists(r_dir):\n",
    "    os.mkdir(r_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c642715-928e-41ee-a91d-14e0d14fb5bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "from wavelet_pseudotime.load_data import load_slavov, load_slavov_xomic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b4114f7-c9b2-4e3c-8e86-927499e2aa83",
   "metadata": {},
   "outputs": [],
   "source": [
    "p, t = wavelet_pseudotime.load_data.load_slavov_xomic()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbaa9847-e7ee-4e14-9213-b321bbdb61fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "p"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1926de0c-040f-45b7-9a60-0b5c47510294",
   "metadata": {},
   "outputs": [],
   "source": [
    "waves_p, scores_p, ps_p, adata_p = wavelet_pseudotime.process.pipeline_slavov(p, save_name=f\"{r_dir}/slavov_p.hd5a\", repeat=True, exclude_pt_ends=(0, 0.975))\n",
    "waves_t, scores_t, ps_t, adata_t = wavelet_pseudotime.process.pipeline_slavov(t, save_name=f\"{r_dir}/slavov_t.hd5a\", repeat=True, exclude_pt_ends=(0,1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c494c019-0e58-4ece-a125-4a8c038add91",
   "metadata": {},
   "outputs": [],
   "source": [
    "# waves_p, scores_p, ps_p, adata_p = wavelet_pseudotime.process.pipeline_slavov(p, save_name=\"slavov_p.hd5a\", repeat=True, exclude_pt_ends=(0, 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37f45eca-3b7b-401b-b137-1d6dd3ff5c4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "panel_6a = adata_t.obs[\"dpt_pseudotime\"].copy()\n",
    "panel_6b = adata_p.obs[\"dpt_pseudotime\"].copy()\n",
    "\n",
    "\n",
    "plt.hist(adata_t.obs[\"dpt_pseudotime\"], bins=100)\n",
    "plt.xlabel(\"Pseudotime\")\n",
    "plt.ylabel(\"Frequency\")\n",
    "plt.title(\"Distribution of Cells in Pseudotime (scRNA)\")\n",
    "# plt.savefig(f\"{r_dir}/pt_dist_scrna.png\")\n",
    "\n",
    "plt.figure()\n",
    "plt.hist(adata_p.obs[\"dpt_pseudotime\"], bins=100)\n",
    "plt.xlabel(\"Pseudotime\")\n",
    "plt.ylabel(\"Frequency\")\n",
    "plt.title(\"Distribution of Cells in Pseudotime (proteins)\")\n",
    "# plt.savefig(f\"{r_dir}/pt_dist_prot.png\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f40591a7-65b7-420f-bbcf-701dfe31f457",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_p.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae03a144-104f-4d2c-813e-e48c87d25b35",
   "metadata": {},
   "outputs": [],
   "source": [
    "prot = wavelet_pseudotime.load_data.load_slavov()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4ec97db-6686-43a1-9f00-1bdb9ef4c945",
   "metadata": {},
   "outputs": [],
   "source": [
    "prot.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88c21272-9c49-4691-a10b-924c7c2cbd1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_p.obs[\"celltype\"] = prot.obs[\"celltype\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1af29430-0d9b-407c-a02b-043497e707d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure they match\n",
    "for idx in adata_p.obs.index:\n",
    "    if adata_p.obs.loc[idx, \"celltype\"] != adata_p.obs.loc[idx, \"celltype\"]:\n",
    "        print(idx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2bc51ed-228a-4c0e-adc6-25888f1e1bf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "g = \"LYZ\"\n",
    "m0_cells = (adata_p.obs[\"celltype\"] == \"sc_m0\").values\n",
    "u_cells = ~m0_cells\n",
    "p07 = (adata_p.obs[\"dpt_pseudotime\"] < 1)\n",
    "plt.plot(adata_p.obs.loc[m0_cells & p07, \"dpt_pseudotime\"], adata_p[m0_cells & p07, g].X[:, 0], \".\", label=\"m0\")\n",
    "plt.plot(adata_p.obs.loc[u_cells & p07, \"dpt_pseudotime\"], adata_p[u_cells & p07, g].X[:, 0], \".\", label=\"u\")\n",
    "plt.legend()\n",
    "plt.xlabel(\"Pseudotime\")\n",
    "plt.ylabel(\"Protein Quant.\")\n",
    "plt.title(f\"Expression of {g} along pseudotime\")\n",
    "# plt.savefig(f\"{r_dir}/example_dist.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e48549b-e22d-44ec-99a4-0d26f1afd8d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Define number of bins for pseudotime\n",
    "num_bins = 20\n",
    "\n",
    "# Bin psupertime into intervals\n",
    "adata_p.obs['pt_bin'] = pd.qcut(adata_p.obs['dpt_pseudotime'], q=num_bins, labels=False)\n",
    "\n",
    "# Compute proportion of cells in each phase per bin\n",
    "phase_proportions = adata_p.obs.groupby(['pt_bin', 'celltype']).size().unstack(fill_value=0)\n",
    "phase_proportions = phase_proportions.div(phase_proportions.sum(axis=1), axis=0)\n",
    "panel_6c = phase_proportions.copy()\n",
    "# Create stacked area plot\n",
    "plt.figure(figsize=(8, 5))\n",
    "plt.stackplot(phase_proportions.index, phase_proportions.T.values, labels=phase_proportions.columns, alpha=0.8)\n",
    "\n",
    "# Formatting\n",
    "plt.xlabel(\"Pseudotime Bin\")\n",
    "plt.ylabel(\"Proportion of Cells\")\n",
    "plt.xticks(ticks=[0, (num_bins-1)//2, num_bins-1], labels=[0,0.5, 1])\n",
    "plt.title(\"Cell Type Proportions Across Pseudotime\")\n",
    "plt.legend(title=\"Phase\", bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.grid(False)\n",
    "\n",
    "# Show plot\n",
    "# plt.savefig(f\"{r_dir}/cell_prop.png\", bbox_inches=\"tight\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdeb2849-ca43-46b3-a594-a7d90ea3b1b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Bin the xlocation values into a specified number of bins\n",
    "num_bins = 20\n",
    "adata_p.obs['pt_bin'] = pd.cut(adata_p.obs['dpt_pseudotime'], bins=num_bins)\n",
    "\n",
    "# Create a crosstab to get counts for each bin and class_label, then normalize each bin to get fractions\n",
    "ct = pd.crosstab(adata_p.obs['pt_bin'], adata_p.obs['celltype'])\n",
    "fractions = ct.div(ct.sum(axis=1), axis=0)\n",
    "\n",
    "# Get the midpoints of each bin for the x-axis.\n",
    "bin_midpoints = np.array([(interval.left + interval.right) / 2 for interval in fractions.index])\n",
    "\n",
    "# Prepare data for the stackplot.\n",
    "categories = fractions.columns.tolist()\n",
    "stack_data = [fractions[cat].values for cat in categories]\n",
    "\n",
    "# Create the stackplot\n",
    "fig, ax = plt.subplots()#figsize=(5, 5))\n",
    "fig6c_x = bin_midpoints.copy()\n",
    "fig6c_y = stack_data.copy()\n",
    "fig6c_labels = categories.copy()\n",
    "ax.stackplot(bin_midpoints, stack_data, labels=categories)\n",
    "\n",
    "# Customize the plot: labels, title, legend.\n",
    "ax.set_xlabel('xlocation (bin midpoints)')\n",
    "ax.set_ylabel('Fraction')\n",
    "ax.set_title('Stacked Fraction Distribution of Class Labels by Binned xlocation')\n",
    "# ax.legend(loc='upper right')\n",
    "\n",
    "# Remove extra whitespace/margins around the plot.\n",
    "# Set x limits exactly to the first and last bin midpoints.\n",
    "ax.set_xlim(bin_midpoints[0], bin_midpoints[-1])\n",
    "# Remove all margins from both axes.\n",
    "ax.margins(x=0, y=0)\n",
    "# Adjust subplot parameters to use all the figure area.\n",
    "plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "\n",
    "plt.savefig(f\"{r_dir}/cell_prop.png\", bbox_inches=\"tight\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d96b4ef8-b1c7-4bee-95f3-4f50afdbc997",
   "metadata": {},
   "outputs": [],
   "source": [
    "idx = np.where(phase_proportions[\"sc_m0\"] > 0.5)[-1][0]\n",
    "fifty_trans = idx / phase_proportions.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7df8d5e-3773-4710-9378-ba3a09ecebb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "fifty_trans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e82b4f1c-ff47-40a8-87ad-c012b9c64378",
   "metadata": {},
   "outputs": [],
   "source": [
    "thresh = 1.2\n",
    "g_above_thresh = [k for k, v in scores_p.items() if v > thresh]\n",
    "print(len(g_above_thresh))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1993fdf-eb81-4cb2-b333-94320300b2a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open(f\"{r_dir}/genes_above_threshold.txt\", \"w\")\n",
    "for g in g_above_thresh:\n",
    "    f.write(f\"{g}\\n\")\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0568ce8c-ef6e-4fb1-934b-8a3994a01256",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clustered pseudotimecourses\n",
    "\n",
    "pt_above_thresh = []\n",
    "for g in g_above_thresh:\n",
    "    pt_above_thresh.append(ps_p[g])\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.metrics import silhouette_score\n",
    "import numpy as np\n",
    "from collections import defaultdict as dd\n",
    "kmeans = KMeans(n_clusters=4, random_state=0)\n",
    "labels = kmeans.fit_predict(pt_above_thresh)\n",
    "c = dd(list)\n",
    "for idx_g, g in enumerate(g_above_thresh):\n",
    "    c[labels[idx_g]].append(g)\n",
    "fig, axs = plt.subplots(2,2)\n",
    "\n",
    "for idx in range(len(c)):\n",
    "    ax = axs[np.unravel_index(idx, axs.shape)]\n",
    "    for g in c[idx]:\n",
    "        x = np.linspace(np.min(bin_midpoints), np.max(bin_midpoints), len(ps_p[g]))\n",
    "        ax.plot(x, ps_p[g])\n",
    "    lin_loc = (np.max(bin_midpoints) - np.min(bin_midpoints)) * fifty_trans + np.min(bin_midpoints)\n",
    "    ax.axvline(lin_loc, linestyle=\"--\", label=\"50% transition\")\n",
    "    ax.set_title(f\"Cluster {idx}\")\n",
    "    if idx == 0:\n",
    "        ax.legend()\n",
    "    if idx > 1:\n",
    "        ax.set_xlabel(\"Pseudotime\")\n",
    "    if idx % 2 == 0:\n",
    "        ax.set_ylabel(\"Gene expression\")\n",
    "fig.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba3fb275-a2d8-4f29-8fd3-4672c2b3d923",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mean timecourses\n",
    "\n",
    "x = np.zeros(len(ps_p[c[0][0]]))\n",
    "count = 0\n",
    "kk = list(c.keys())\n",
    "kk.sort()\n",
    "# for idx in c.keys():\n",
    "for idx in kk:\n",
    "    count = 0\n",
    "    x = np.zeros(len(ps_p[c[0][0]]))\n",
    "    # print(len(c[idx]))  # number of genes in each cluster\n",
    "    for g in c[idx]:\n",
    "        x += ps_p[g]\n",
    "        count += 1\n",
    "    x /= count\n",
    "    minx = np.min(bin_midpoints)\n",
    "    maxx = np.max(bin_midpoints)\n",
    "    num = len(x)\n",
    "    plt.plot(np.linspace(minx, maxx, num), x, label=f\"Cluster {idx}\")\n",
    "lin_pos = (maxx - minx) * fifty_trans + minx\n",
    "plt.axvline(lin_pos, linestyle=\"--\", label=\"50% transition\")\n",
    "plt.legend()\n",
    "plt.title(\"Mean pseudotimecourses\")\n",
    "plt.xlabel(\"Pseudotime\")\n",
    "plt.ylabel(\"Protein quant.\")\n",
    "plt.savefig(f\"{r_dir}/fig7_cluster_mean_timecourses.png\")\n",
    "plt.savefig(f\"{r_dir}/fig7_cluster_mean_timecourses.svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07e98b7e-b33c-494c-99a9-94a9d71ce5a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(8, 6))\n",
    "gs = gridspec.GridSpec(3, 2, figure=fig)\n",
    "ax = fig.add_subplot(gs[0, :])\n",
    "c = dd(list)\n",
    "for idx_g, g in enumerate(g_above_thresh):\n",
    "    c[labels[idx_g]].append(g)\n",
    "x = np.zeros(len(ps_p[c[0][0]]))\n",
    "count = 0\n",
    "kk = list(c.keys())\n",
    "kk.sort()\n",
    "# for idx in c.keys():\n",
    "for idx in kk:\n",
    "    count = 0\n",
    "    x = np.zeros(len(ps_p[c[0][0]]))\n",
    "    for g in c[idx]:\n",
    "        x += ps_p[g]\n",
    "        count += 1\n",
    "    x /= count\n",
    "    minx = np.min(bin_midpoints)\n",
    "    maxx = np.max(bin_midpoints)\n",
    "    num = len(x)\n",
    "    ax.plot(np.linspace(minx, maxx, num), x, label=f\"Cluster {idx}\")\n",
    "lin_pos = (maxx - minx) * fifty_trans + minx\n",
    "# plt.axvline(fifty_trans * len(ps_p[g]), linestyle=\"--\", label=\"50% transition\")\n",
    "ax.axvline(lin_pos, linestyle=\"--\", label=\"50% transition\")\n",
    "plt.legend(bbox_to_anchor=(1,1))\n",
    "ax.set_title(\"Mean pseudotimecourses\", fontsize=16)\n",
    "ax.set_xlabel(\"Pseudotime\", fontsize=16)\n",
    "ax.set_ylabel(\"Protein quant.\", fontsize=16)\n",
    "\n",
    "pt_above_thresh = []\n",
    "for g in g_above_thresh:\n",
    "    pt_above_thresh.append(ps_p[g])\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.metrics import silhouette_score\n",
    "import numpy as np\n",
    "from collections import defaultdict as dd\n",
    "kmeans = KMeans(n_clusters=4, random_state=0)\n",
    "labels = kmeans.fit_predict(pt_above_thresh)\n",
    "\n",
    "gs_list = [gs[1,0], gs[1,1], gs[2,0], gs[2,1]]\n",
    "\n",
    "for idx in range(len(c)):\n",
    "    gs0 = gs_list[idx]\n",
    "    ax = fig.add_subplot(gs0)\n",
    "    for g in c[idx]:\n",
    "        x = np.linspace(np.min(bin_midpoints), np.max(bin_midpoints), len(ps_p[g]))\n",
    "        ax.plot(x, ps_p[g])\n",
    "    lin_loc = (np.max(bin_midpoints) - np.min(bin_midpoints)) * fifty_trans + np.min(bin_midpoints)\n",
    "    # ax.axvline(fifty_trans * len(ps_p[g]), linestyle=\"--\", label=\"50% transition\")\n",
    "    ax.axvline(lin_loc, linestyle=\"--\", label=\"50% transition\")\n",
    "    # ax.axvline(last_g1, linestyle=\"--\")\n",
    "    # ax.axvline(last_s, linestyle=\"--\")\n",
    "    ax.set_title(f\"Cluster {idx}\", fontsize=16)\n",
    "    if idx == 0:\n",
    "        ax.legend()\n",
    "    if idx > 1:\n",
    "        ax.set_xlabel(\"Pseudotime\", fontsize=16)\n",
    "    if idx % 2 == 0:\n",
    "        ax.set_ylabel(\"Protein quant.\", fontsize=16)\n",
    "fig.tight_layout()\n",
    "\n",
    "plt.savefig(f\"{r_dir}/fig7_cluster_mean_timecourses.png\")\n",
    "plt.savefig(f\"{r_dir}/fig7_cluster_mean_timecourses.svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "503d9512-3a65-4ed9-8470-1b194810d53e",
   "metadata": {},
   "outputs": [],
   "source": [
    "wt = wavelet_pseudotime.wavelets.WaveletTransform(scales=[1,2,3,4], wavelet=\"mexh\")\n",
    "coef, _ = wt.apply(x)\n",
    "plt.imshow(coef)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3eeae401-bffc-486d-9a96-d954533cffef",
   "metadata": {},
   "source": [
    "# Figure assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1535f66-79ee-4017-a034-e13c150384a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from string import ascii_uppercase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e32711a-20e4-4240-af97-26a38c6c21b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(10, 12))\n",
    "gs = gridspec.GridSpec(5, 2, figure=fig)\n",
    "fontsize=16\n",
    "\n",
    "ax = fig.add_subplot(gs[0,0])\n",
    "ax.hist(panel_6a, bins=100);\n",
    "ax.set_ylabel(\"Frequency\", fontsize=fontsize)\n",
    "ax.set_title(\"scRNA pseudotime\", fontsize=fontsize)\n",
    "ax.text(0.02, 0.95, ascii_uppercase[0], transform=ax.transAxes,\n",
    "            fontsize=16, fontweight='bold', va='top', ha='left')\n",
    "\n",
    "ax = fig.add_subplot(gs[0,1])\n",
    "ax.hist(panel_6b, bins=100);\n",
    "ax.set_title(\"scP pseudotime\", fontsize=fontsize)\n",
    "ax.set_xlim([0,1])\n",
    "ax.text(0.02, 0.95, ascii_uppercase[1], transform=ax.transAxes,\n",
    "            fontsize=16, fontweight='bold', va='top', ha='left')\n",
    "\n",
    "ax = fig.add_subplot(gs[1,0])\n",
    "ax.stackplot(fig6c_x, fig6c_y, labels=fig6c_labels)\n",
    "ax.set_xlabel('Pseudotime', fontsize=fontsize)\n",
    "ax.set_ylabel('Fraction', fontsize=fontsize)\n",
    "ax.set_title('scP cell type fraction', fontsize=fontsize)\n",
    "ax.text(0.02, 0.95, ascii_uppercase[2], transform=ax.transAxes,\n",
    "            fontsize=16, fontweight='bold', va='top', ha='left')\n",
    "\n",
    "# Remove extra whitespace/margins around the plot.\n",
    "# Set x limits exactly to the first and last bin midpoints.\n",
    "ax.set_xlim(bin_midpoints[0], bin_midpoints[-1])\n",
    "# Remove all margins from both axes.\n",
    "ax.margins(x=0, y=0)\n",
    "# Adjust subplot parameters to use all the figure area.\n",
    "plt.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "\n",
    "# g = \"LMNA\"\n",
    "g = \"CAPG\"\n",
    "# ax = axs[1,1]\n",
    "ax = fig.add_subplot(gs[1,1])\n",
    "m0_cells = (adata_p.obs[\"celltype\"] == \"sc_m0\").values\n",
    "u_cells = ~m0_cells\n",
    "p07 = (adata_p.obs[\"dpt_pseudotime\"] < 0.7)\n",
    "ax.plot(adata_p.obs.loc[m0_cells & p07, \"dpt_pseudotime\"], adata_p[m0_cells & p07, g].X[:, 0], \".\", label=\"m0\")\n",
    "ax.plot(adata_p.obs.loc[u_cells & p07, \"dpt_pseudotime\"], adata_p[u_cells & p07, g].X[:, 0], \".\", label=\"u\")\n",
    "ax.legend()\n",
    "ax.set_xlabel(\"Pseudotime\", fontsize=fontsize)\n",
    "ax.set_ylabel(\"Protein Quant.\", fontsize=fontsize)\n",
    "ax.set_title(f\"Expression of {g}\", fontsize=fontsize)\n",
    "ax.plot(np.linspace(0,np.max(adata_p.obs[\"dpt_pseudotime\"]), len(ps_p[g])), ps_p[g])\n",
    "ax.text(0.02, 0.15, ascii_uppercase[3], transform=ax.transAxes,\n",
    "            fontsize=16, fontweight='bold', va='top', ha='left')\n",
    "\n",
    "ax = fig.add_subplot(gs[2,:])\n",
    "c = dd(list)\n",
    "for idx_g, g in enumerate(g_above_thresh):\n",
    "    c[labels[idx_g]].append(g)\n",
    "x = np.zeros(len(ps_p[c[0][0]]))\n",
    "count = 0\n",
    "kk = list(c.keys())\n",
    "kk.sort()\n",
    "for idx in kk:\n",
    "    count = 0\n",
    "    x = np.zeros(len(ps_p[c[0][0]]))\n",
    "    print(len(c[idx]))\n",
    "    for g in c[idx]:\n",
    "        x += ps_p[g]\n",
    "        count += 1\n",
    "    x /= count\n",
    "    minx = np.min(bin_midpoints)\n",
    "    maxx = np.max(bin_midpoints)\n",
    "    num = len(x)\n",
    "    ax.plot(np.linspace(minx, maxx, num), x, label=f\"Cluster {idx}\")\n",
    "lin_pos = (maxx - minx) * fifty_trans + minx\n",
    "ax.axvline(lin_pos, linestyle=\"--\", label=\"50% transition\")\n",
    "plt.legend(bbox_to_anchor=(1,1))\n",
    "ax.set_title(\"Mean pseudotimecourses\", fontsize=16)\n",
    "ax.set_xlabel(\"Pseudotime\", fontsize=16)\n",
    "ax.set_ylabel(\"Protein quant.\", fontsize=16)\n",
    "ax.text(0.02, 0.95, ascii_uppercase[4], transform=ax.transAxes,\n",
    "            fontsize=16, fontweight='bold', va='top', ha='left')\n",
    "\n",
    "\n",
    "# pseudotimecourse\n",
    "pt_above_thresh = []\n",
    "for g in g_above_thresh:\n",
    "    pt_above_thresh.append(ps_p[g])\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.metrics import silhouette_score\n",
    "import numpy as np\n",
    "from collections import defaultdict as dd\n",
    "kmeans = KMeans(n_clusters=4, random_state=0)\n",
    "labels = kmeans.fit_predict(pt_above_thresh)\n",
    "\n",
    "gs_list = [gs[3,0], gs[3,1], gs[4,0], gs[4,1]]\n",
    "\n",
    "for idx in range(len(c)):\n",
    "    gs0 = gs_list[idx]\n",
    "    ax = fig.add_subplot(gs0)\n",
    "    for g in c[idx]:\n",
    "        x = np.linspace(np.min(bin_midpoints), np.max(bin_midpoints), len(ps_p[g]))\n",
    "        ax.plot(x, ps_p[g])\n",
    "    lin_loc = (np.max(bin_midpoints) - np.min(bin_midpoints)) * fifty_trans + np.min(bin_midpoints)\n",
    "    ax.axvline(lin_loc, linestyle=\"--\", label=\"50% transition\")\n",
    "    ax.set_title(f\"Cluster {idx}\", fontsize=16)\n",
    "    if idx == 0:\n",
    "        ax.legend()\n",
    "    if idx > 1:\n",
    "        ax.set_xlabel(\"Pseudotime\", fontsize=16)\n",
    "    if idx % 2 == 0:\n",
    "        ax.set_ylabel(\"Protein quant.\", fontsize=16)\n",
    "    ax.text(0.02, 0.95, ascii_uppercase[5+idx], transform=ax.transAxes,\n",
    "            fontsize=16, fontweight='bold', va='top', ha='left')\n",
    "\n",
    "fig.tight_layout()\n",
    "plt.savefig(f\"{r_dir}/fig6_slavov.png\")\n",
    "plt.savefig(f\"{r_dir}/fig6_slavov.svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a32a05f-1256-43e4-8963-7ad02692c860",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd77ba9c-903e-4b01-86bd-089e25f5ae69",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(adata_p, groupby=\"celltype\")\n",
    "top100 = adata_p.uns[\"rank_genes_groups\"][\"names\"][:100]\n",
    "adata_p.obs[\"celltype\"].cat.categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "239322a3-276f-4968-85b1-6e7b5e38d1af",
   "metadata": {},
   "outputs": [],
   "source": [
    "top100_m0 = []\n",
    "top100_u = []\n",
    "for c in top100:\n",
    "    top100_m0.append(c[0])\n",
    "    top100_u.append(c[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b86760f-df72-46d7-9f2a-fa1f4d37e9ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "score_list = []\n",
    "for k, v in scores_p.items():\n",
    "    score_list.append({\"protein\": k, \"score\": v})\n",
    "sorted_scores = sorted(score_list, key=lambda x: x[\"score\"])\n",
    "top100_proteins = [k[\"protein\"] for k in sorted_scores[-100:]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d53cac7-da5c-44aa-b60e-4fca0a1a4092",
   "metadata": {},
   "outputs": [],
   "source": [
    "m0_overlap = set(top100_m0).intersection(set(top100_proteins))\n",
    "u_overlap = set(top100_u).intersection(set(top100_proteins))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7df99c34-670e-4c35-a390-2c4d130f2fb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(m0_overlap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45fc09ad-2955-4080-a02b-e680e856186d",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(u_overlap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "548d8eff-d241-4fb2-936e-f579e8fcbdc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.patches import Circle\n",
    "\n",
    "\n",
    "def draw_venn_diagram(group1_count: int,\n",
    "                      group2_count: int,\n",
    "                      group3_count: int,\n",
    "                      overlap12_count: int,\n",
    "                      overlap23_count: int) -> None:\n",
    "    # create a new figure and axis\n",
    "    fig, ax = plt.subplots(figsize=(7, 6))\n",
    "\n",
    "    # Define circle properties.\n",
    "    radius = 1.0\n",
    "    circle1_center = (0, 0)  # Group 1\n",
    "    circle2_center = (1.4, 0)  # Group 2 overlaps with both Group 1 and Group 3\n",
    "    circle3_center = (2.8, 0)  # Group 3\n",
    "    group1_label = \"Monocyte DE\"\n",
    "    group2_label = \"TE\"\n",
    "    group3_label = \"Macrophage DE\"\n",
    "    fontsize=16\n",
    "    # Create circles using Circle patch.\n",
    "    circle1 = Circle(circle1_center, radius, color='red', alpha=0.4, label=group1_label)\n",
    "    circle2 = Circle(circle2_center, radius, color='green', alpha=0.4, label=group2_label)\n",
    "    circle3 = Circle(circle3_center, radius, color='blue', alpha=0.4, label=group3_label)\n",
    "\n",
    "    # Add circles to the plot.\n",
    "    for circle in (circle1, circle2, circle3):\n",
    "        ax.add_patch(circle)\n",
    "\n",
    "    # Annotate each circle with the group total count.\n",
    "    # For Group 1, place the number in the left (exclusive) region.\n",
    "    ax.text(circle1_center[0] - 0.5, circle1_center[1] + 0.1, f\"{group1_count}\",\n",
    "            fontsize=fontsize, color='darkred', weight='bold')\n",
    "\n",
    "    # For Group 2, place the main count in the upper region.\n",
    "    ax.text(circle2_center[0] - 0.2, circle2_center[1] + 0.4, f\"{group2_count}\",\n",
    "            fontsize=fontsize, color='darkgreen', weight='bold')\n",
    "\n",
    "    # For Group 3, place the number in the right (exclusive) region.\n",
    "    ax.text(circle3_center[0] + 0.4, circle3_center[1] + 0.1, f\"{group3_count}\",\n",
    "            fontsize=fontsize, color='darkblue', weight='bold')\n",
    "\n",
    "    # Annotate overlaps:\n",
    "    # Overlap between Group 1 and Group 2:\n",
    "    # Compute an approximate midpoint between the centers of circle1 and circle2.\n",
    "    overlap_12_pos = ((circle1_center[0] + circle2_center[0]) / 2, (circle1_center[1] + circle2_center[1]) / 2 - 0.2)\n",
    "    ax.text(overlap_12_pos[0] - 0.1, overlap_12_pos[1], f\"{overlap12_count}\",\n",
    "            fontsize=fontsize, color='black', weight='bold')\n",
    "\n",
    "    # Overlap between Group 2 and Group 3:\n",
    "    overlap_23_pos = ((circle2_center[0] + circle3_center[0]) / 2, (circle2_center[1] + circle3_center[1]) / 2 - 0.2)\n",
    "    ax.text(overlap_23_pos[0] - 0.1, overlap_23_pos[1], f\"{overlap23_count}\",\n",
    "            fontsize=fontsize, color='black', weight='bold')\n",
    "\n",
    "    # (Optional) Add group name labels below the circles.\n",
    "    \n",
    "    ax.text(circle1_center[0] - 0.5, circle1_center[1] - 1.2, group1_label, fontsize=fontsize, color='darkred')\n",
    "    ax.text(circle2_center[0] - 0.1, circle2_center[1] - 1.2, group2_label, fontsize=fontsize, color='darkgreen')\n",
    "    ax.text(circle3_center[0] - 0.5, circle3_center[1] - 1.2, group3_label, fontsize=fontsize, color='darkblue')\n",
    "\n",
    "    # Adjust plot limits so all circles are clearly visible.\n",
    "    ax.set_xlim(-1, 4)\n",
    "    ax.set_ylim(-1, 1)\n",
    "\n",
    "    # Use an equal aspect ratio for proper circle representation.\n",
    "    ax.set_aspect('equal')\n",
    "    ax.axis('off')\n",
    "\n",
    "    # Add a title and show the plot.\n",
    "    # plt.title('Venn Diagram with Group Totals and Overlaps')\n",
    "    fig.subplots_adjust(left=0, right=1, top=1, bottom=0)\n",
    "    plt.title(\"Overlap between DE genes and TE genes\", fontsize=20)\n",
    "    fig.tight_layout()\n",
    "    # plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d08a7765-ffe4-46dd-9740-e7181a5f22a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "draw_venn_diagram(len(top100_u)-len(u_overlap), len(top100_proteins)-len(u_overlap)-len(m0_overlap), len(top100_m0)-len(m0_overlap), len(u_overlap), len(m0_overlap))\n",
    "plt.savefig(f\"{r_dir}/venn_slavov.png\", bbox_inches=\"tight\")\n",
    "plt.savefig(f\"{r_dir}/venn_slavov.svg\", bbox_inches=\"tight\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python311-wavelets",
   "language": "python",
   "name": "python311-wavelets"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
